File: 2483.minimum-penalty-for-a-shop.go
/*
 * @lc app=leetcode id=2483 lang=golang
 *
 * [2483] Minimum Penalty for a Shop
 */

// @lc code=start
func bestClosingTime(customers string) int {
    right_y := make([]int, len(customers)+1);
	cnt := 0;
	for i:=len(customers)-1;i>=0;i--{
		if customers[i] == 'Y' {
			cnt++;
		}
		right_y[i] = cnt;
	}

	res := -1;
	minimum := math.MaxInt;
	cnt = 0;
	for i:=0;i<=len(customers);i++{
		penalty := cnt + right_y[i];
		if minimum > penalty {
			minimum = penalty;
			res = i;
		}
		if i<len(customers) && customers[i] == 'N'{
			cnt++;
		}
	}
	return res;
}
// @lc code=end

File: 2483.minimum-penalty-for-a-shop.cpp
/*
 * @lc app=leetcode id=2483 lang=cpp
 *
 * [2483] Minimum Penalty for a Shop
 */

// @lc code=start
class Solution {
public:
    // dp
    int bestClosingTime(string customers) {
        vector<int> right_y_cnt(customers.size()+1, 0);
        int cnt = 0;
        for(int i=customers.size()-1;i>=0;i--){
            cnt += customers[i] == 'Y';
            right_y_cnt[i] = cnt;
        }

        int minimum = INT_MAX;
        int res = -1;
        cnt = 0;
        for(int i=0;i<=customers.size();i++){
            int penalty = right_y_cnt[i] + cnt;
            if(minimum > penalty){
                minimum = penalty;
                res = i;
            }
            cnt += customers[i] == 'N';
        }
        return res;
    }
};
// @lc code=end

